lib/src/folding/foldable_block_matcher.dart

```diff
@@ -0,0 +1,91 @@
+import 'dart:math';
+
+import '../code/code_line.dart';
+import 'foldable_block.dart';
+
+/// Matches foldable blocks before and after an edit to preserve their
+/// folding state.
+///
+/// Blocks match if they go in the same order and have the same content.
+class FoldableBlockMatcher {
+  final List<CodeLine> oldLines;
+  final List<FoldableBlock> newBlocks;
+  final List<CodeLine> newLines;
+  final oldToNew = <FoldableBlock, FoldableBlock>{};
+  final newFoldedBlocks = <FoldableBlock>{};
+
+  FoldableBlockMatcher({
```

Comment:
I don't see an advantage of a static method here.

Reply:
We may need to expose additional fields in the future like the mapping between the blocks. This is why I use a class.

Link 1 is about testing and mocking. We enclose the whole object life cycle in a test so we are not impacted by this.
Link 2 is not applicable because it is about lazy logic, but this logic is always needed for the task so it cannot be lazy.